graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -195,90 +231,24 @@ public String getFingerprint() {
         return matches;
     }
 
-    private void matchRules(Message message, Set<String> fields, Map<String, List<Rule>> rules, Map<Stream, StreamMatch> matches) {
-        for (String field : fields) {
-            for (Rule rule : rules.get(field)) {
-                registerMatch(matches, rule.match(message));
-            }
-        }
-    }
-
-    private void matchRulesWithTimeout(final Message message, Set<String> fields, Map<String, List<Rule>> rules, Map<Stream, StreamMatch> matches, Set<Stream> timeouts) {
-        for (String field : fields) {
-            for (final Rule rule : rules.get(field)) {
-                final Callable<Stream> task = new Callable<Stream>() {
-                    @Override
-                    public Stream call() {
-                        return rule.match(message);
-                    }
-                };
-
-                try {
-                    final Stream match = timeLimiter.callWithTimeout(task, streamProcessingTimeout, TimeUnit.MILLISECONDS, true);
-
-                    registerMatch(matches, match);
-                } catch (UncheckedTimeoutException e) {
-                    timeouts.add(rule.getStream());
-                } catch (Exception e) {
-                    LOG.error("Unexpected stream rule exception.", e);
-                }
-            }
-        }
-    }
-
-    private void registerMatch(Map<Stream, StreamMatch> matches, Stream match) {
-        if (match != null) {
-            if (!matches.containsKey(match)) {
-                matches.put(match, new StreamMatch(match));
-            }
-            matches.get(match).increment();
-        }
-    }
-
-    private void addRule(Map<String, List<Rule>> rules, Set<String> fields, String field, Rule rule) {
-        fields.add(field);
-
-        if (! rules.containsKey(field)) {
-            rules.put(field, Lists.newArrayList(rule));
-        } else {
-            rules.get(field).add(rule);
-        }
-    }
-
-    private class StreamMatch {
-        private final int ruleCount;
-        private int matches = 0;
-        private final Stream stream;
-
-        public StreamMatch(Stream stream) {
-            this.stream = stream;
-            this.ruleCount = stream.getStreamRules().size();
-        }
-
-        public void increment() {
-            matches++;
-        }
-
-        public boolean isMatched() {
-            switch (stream.getMatchingType()) {
-                case AND: return ruleCount == matches;
-                case OR: return ruleCount > 0;
-                default: return ruleCount == matches;
-            }
-        }
-    }
-
-    private class Rule {
+    protected class Rule {
         private final Stream stream;
         private final StreamRule rule;
         private final StreamRuleMatcher matcher;
+        private final Boolean sufficient;
```

Comment:
While `sufficient` is shorter, I think `shortCircuit` or `shortCircuitRule` describes better, what this flag denotes in reality.